app/ante/msg_gatekeeper.go

```diff
@@ -49,6 +49,12 @@ func (mgk MsgVersioningGateKeeper) hasInvalidMsg(ctx sdk.Context, acceptedMsgs m
 			if err != nil {
 				return err
 			}
+			// App version 2 introduces a new constraint that prevents a MsgExec inside a MsgExec.
+			if ctx.BlockHeader().Version.App >= 2 {
+				if err = mgk.hasNestedMsgExec(ctx, nestedMsgs); err != nil {
+					return err
+				}
+			}
 			if err = mgk.hasInvalidMsg(ctx, acceptedMsgs, nestedMsgs); err != nil {
 				return err
 			}
```

Comment:
Is `mgk.hasInvalidMsg` still needed since the new nested check would catch it before? (unless ofc this was shipped into prod already w/ v1)

Reply:
`hasInvalidMsg` is still necessary to disallow `TryMsgUpgrade` within the `MsgExec`